readme.md
# TAG
###### An attempt to train a neural network to play tag

This is my first attempt at a machine learning project, and as such will most likely end poorly.

### Game Rules

There is a set amount of Hunters and Players within the game.
Hunters get points for being close to players and capturing them (the being near is more for teaching them more easily)
Players get points for the number of seconds they are alive and lose points for being near Hunters
There are walls which must be maneuvered around by both Players and Hunters, and also block the Hunter's ability to see players

### Running Simulation
Go to your terminal of choice and type `python3 sim.py`, then simply enter the desired values and it will either read a previous network or initialize a new one.

### Network
The network is just about as simple as it comes. It's a feed forward network thats being trained with the simplest version of evolution.
The layer count can be changed in the nnwk.py file 


changelog.txt
ok this is turning into a genuine project so I should *probably* have a changelog. here is is

V 1.2 (Hep3astus):
  Actually added the capturing mechanism. The nets were too bad at the game to have it happen but then I realized it was very broken
  Did some other changes that I don't remember


V 1.13 (Heph3astus):
  added a punishment to the hunter for staying still
    This should promote more exploration and stop it from just staying in one place collecting points from players nearby

V 1.12 (Heph3astus):
  changed colour of collision dots
  fixed collision system

V 1.11 (Heph3astus):
  created a bandaid solution for an index out of range error that I'm still not sure the cause of
    Will remove and update once I find the issue

V 1.1 (Heph3astus):

  Redid the collision system for players and hunters to use the raycaster
    the other one was being extremely sensitive and blocking lots of movement
    Also added the ability to not display anything with the raycaster
      otherwise would have shifted doubles of the rays
    removed accidental duplication of execution of functions by players and hunters
        meant that all functions were going twice as much distance as expected

V 1.0 (not really but its when I'm staring this) (By Heph3astus):
  Bug fixes of course

  Added a system to change counts of layers and changed filenames to reflect that
    This should let me and anyone else experiment with what layer counts work best
      Yes I could try to implement NEAT but I don't have the skills for that.
        If someone wants to please do and push it but not to master cuz yea...

  Removed getWeights function as it had no use because I put the writing of the network into the network class

  Got rid of useless template which was just for me to remember how to code it

  Fixed bug that was occasionally throwing permission errors for creating a file
    Added relpath to all file grabs because some computers didn't like it

  Fixed how player fitness is calculated.
    The loss was calculated poorly and was accidentally adding instead of subtracting
      Made less important because its every frame

  Added general options to help hopefully ramp up training speed like easy change to frame count of sim

  Fixed fact that output floats were way higher or lower than the biases
    Changed normal distribution shuffle variance to reflect this
    removed the output bias caps

  Added titles to files cuz people might somehow be confused

  Added more rays to the players and hunters so that they can more easily see things (32 rays vs 16)

  Created readme file

  Added this changelog


test.py
from nnwk import NeuralNetwork

test = NeuralNetwork(2,1,7,player)


sim.py
#coordinating script
import os
import tag

simCount = 0
genCount = 0


while True:
    print("sims per generation: ")
    simCount = input()
    try:
        simCount = int(simCount)
        break
    except:
        print("not an int. please try again")

while True:
    print("generation count (1 for continue until quit): ")
    genCount = input()
    try:
        genCount = int(genCount)
        break
    except:
        print("not an int. please try again")


print("display one sim per generation (y/n): ")
display = input()

if display == "y" :
    display = True
elif display == "n":
    display = False
else:
    print("unrecognized entry. defaulting to no")
    display = False

playerCount = 0
while True:
    print("playerCount: ")
    playerCount = input()
    try:
        playerCount = int(playerCount)
        break
    except:
        print("not an int. please try again")

hunterCount = 0
while True:
    print("hunterCount: ")
    hunterCount = input()
    try:
        hunterCount = int(hunterCount)
        break
    except:
        print("not an int. please try again")

print("frames cap of simulations (default to 1000):")
frameLen = input()
try:
    frameLen = int(frameLen)
except:
    print("unrecognized entry. defaulting to 1000")
    frameLen = 1000

sims = [None] * simCount

hShuffleRate = 1
pShuffleRate = 1

info = "genInfo/" + str(playerCount) + ";" + str(hunterCount) + ".txt"

genIndex = 0
if os.path.isfile(info):
    with open(os.path.relpath(info),'r') as f:
        l = f.readlines()
        genIndex = int(l[1])
else:
    with open(info,'w') as f:
        f.writelines(["generation:\n0"])
g = 0
prevHFitness = [0] * 5
while g < genCount:

    print("generation " + str(genIndex))
    results = [None] * simCount

    sims = [tag.Sim(playerCount,hunterCount,pShuffleRate,hShuffleRate,False,frameLen)] * simCount

    if display:
        sims[0].display = True

    #run each game loop

    while True:

        isDone = True

        for i in range(len(sims)):
            if results[i] == None:
                if sims[i].running == False:
                    results[i] = sims[i].getResults()

                    continue
                sims[i].runSim()
                isDone = False

        if isDone:
            break

    hFitness = 0
    pFitness = 0
    hNetwork = None
    pNetwork = None

    hShuffleRate = 0
    pShuffleRate = 0

    for i in range(len(results)):
        if results[i][0] >= hFitness:
            hFitness = results[i][0]
            hNetwork = results[i][1]
            #So many dumb things ive done here...
            hShuffleRate = 1
        if results[i][2] > pFitness:
            pFitness = results[i][2]
            pNetwork = results[i][3]
            #just doing this so that I dont have to edit my code. I'm very lazy
            pShuffleRate = 1

    print("generation complete.")
    print("best hunter fitness: " + str(hFitness))
    print("best player fitness: " + str(pFitness))

    try:
        hNetwork.write("hunter",hunterCount,playerCount)
        pNetwork.write("player",hunterCount,playerCount)
    except:
        pass

    genIndex+=1
    data = None
    with open(info,'r') as f:
        data = f.readlines()

    data[1] = str(genIndex)
    with open(info,'w') as f:
        f.writelines(data)
    if genCount != 1:
        g+=1

    for i in range(len(prevHFitness)-1):
        prevHFitness[i] = prevHFitness[i+1]
    prevHFitness[len(prevHFitness)-1] = hFitness


    #code to reset network if nothings happening
    new = False
    for p in prevHFitness:
        if p != prevHFitness[0]:
            new = True

    if not new:
        try:
            os.remove(os.path.relpath("hunterNets/"+str(playerCount)+";"+str(hunterCount) + "." + str(hNetwork.layerCount) + ".txt"))
        except:
            os.remove(os.path.relpath("hunterNets/"+str(playerCount)+";"+str(hunterCount) + ".3.txt"))


nnwk.py
from numpy import exp
import numpy as np
import os
import random

class NeuralNetwork():

    outputBiases = []
    weights = None
    outputWeights = []
    layerCount = 3

    def __init__(self, playerCount, hunterCount, inputCount, eType):


        random.seed(os.urandom(5000))

        #getting weight data if existing and if not creating it
        n = eType + "Nets/" + str(playerCount) + ";" + str(hunterCount) + "." + str(self.layerCount) + ".txt"
        net = os.path.relpath(n)

        self.weights = np.zeros((inputCount,self.layerCount,inputCount),dtype=float)

        self.inputCount = inputCount

        if os.path.isfile(net):
            with open(os.path.relpath(net)) as f:
                #read weights
                counter = 0
                data = f.readlines()
                for i in range(0,self.layerCount):
                    for x in range (0,inputCount):
                        for z in range(0,inputCount):
                            self.weights[x,i,z] = float(data[z+(inputCount*x)+(inputCount*inputCount*i)])
                            counter+=1

                counter2 = 0
                for i in range (0,5):
                    self.outputWeights.append([])

                for i in range(counter,len(data)-5):
                    self.outputWeights[counter2].append(float(data[i]))
                    if (i-counter+1)%(inputCount) == 0:
                        counter2+=1
                for i in range(len(data)-5,len(data)):
                    self.outputBiases.append(float(data[i]))
        else:

            for i in range(0,self.layerCount):
                for x in range(0,inputCount):
                    for z in range(0,inputCount):
                        num = 2 * random.random() - 1
                        self.weights[x,i,z] = num

            for i in range(0,5):
                self.outputWeights.append([])
                for x in range(0,inputCount):
                    self.outputWeights[i].append(2*random.random() - 1)

            #range is to number of functions
            for i in range(0,5):
                self.outputBiases.append(0.5 * random.random()+0.5)

            #creating and writing file
            newData = []
            for i in range(0,self.layerCount):
                for x in range(0,self.inputCount):
                    for z in range(0,self.inputCount):
                        newData.append(self.weights[x,i,z])
            for i in range(0,5):
                for x in range(0,self.inputCount):
                    newData.append(float(self.outputWeights[i][x]))

            for i in range(0,5):
                newData.append(float(self.outputBiases[i]))
            with open(os.path.relpath(net), 'w+') as f:

                for i in range(0,len(newData)-1):
                    f.write("%s\n" % newData[i])
                f.write("%s" % newData[len(newData)-1])



    def shuffle(self,rate):
        for i in range(0,self.layerCount):
            for x in range(0,self.inputCount):
                for z in range(0,self.inputCount):
                    self.weights[x,i,z] += float(np.random.normal(loc = 0, scale = rate*0.0001,size = 1))
                    if self.weights[x,i,x] > 1.2:
                        self.weights[i,i,x] = 1.2
                    elif self.weights[x,i,x] < -1.2:
                        self.weights[i,i,x] = -1.2
        for i in range(0,5):
            for x in range(0,self.inputCount):
                self.outputWeights[i][x] += float(np.random.normal(loc = 0,scale = rate*0.0001,size = 1))
                if self.outputWeights[i][x] > 1.2:
                    self.outputWeights[i][x] = 1.2
                elif self.outputWeights[i][x] < -1.2:
                    self.outputWeights[i][x] = -1.2
        for i in range(0,5):
            self.outputBiases[i] += float(np.random.normal(loc = 0, scale = rate*0.1,size = 1))

    def write(self,eType,hCount,pCount):
        n = eType + "Nets/" + str(pCount) + ";" + str(hCount) + "." + str(self.layerCount) + ".txt"
        net = os.path.relpath(n)
        #creating and writing file
        newData = []
        for i in range(0,self.layerCount):
            for x in range(0,self.inputCount):
                for z in range(0,self.inputCount):
                    newData.append(self.weights[x,i,z])
        for i in range(0,5):
            for x in range(0,self.inputCount):
                newData.append(float(self.outputWeights[i][x]))

        for i in range(0,5):
            newData.append(float(self.outputBiases[i]))
        with open(net, 'w') as f:

            for i in range(0,len(newData)-1):
                f.write("%s\n" % newData[i])
            f.write("%s" % newData[len(newData)-1])


    def think(self,inputs,entity):
        #variable storing values of previous layer
        previousValues = inputs
        values = previousValues
        for i in self.weights:
            for index, x in enumerate(i):
                val = 0
                div = 0
                for ind, z in enumerate(x):
                    val += float(previousValues[ind]*z)
                    div += float(z)
                values[index] = val
            previousValues = values

        #gets final output layer and turns it into a boolean array to do functions on
        outputs = [False,False,False,False,False]
        for ind, i in enumerate(self.outputWeights):
            if ind == 5:
                break
            val = 0
            for index, x in enumerate(previousValues):
                val += x*i[index]

            if val >= self.outputBiases[ind]:
                outputs[ind] = True

        return outputs


player.py
import random
import pygame
import math
import os
from raycaster import distances

class Player:
    #values for input into the network
    x = 0
    y = 0
    closestHoleX = 0
    closestHoleY = 0
    isInHole = False

    walls = None
    traces = [0] * 16

    rect = None


    captured = False

    fitness = 2000
    loss = 0

    #### IMPORTANT NOTICE. Since I'm too lazy to figure out how to do variable
    #### input node count so that the number of hunters and players can be
    #### changed, instead I am saving weight data to a file which is named the
    #### number of players , the number of hunters.
    #### THEREFORE WHENEVER THE NUMBER OF PLAYERS OR HUNTERS IS CHANGED
    #### IT WILL START BLANK AND WILL HAVE TO COMPLETELY RELEARN EVERYTHING
    #### side note this is actually probably not a bad idea because I can then
    #### experiment with different counts and they may develop completely
    #### different strategies

    #### note to self: fitness for the player will be measured in 50*seconds alive


    speed = 3
    visRad = 50

    def __init__(self,walls,hunterCount,playerCount,seed):
        self.walls = walls
        random.seed(seed)
        x = 0
        y = 0
        rect = None
        good = False
        while not good:
            x = random.randrange(850)+75
            y = random.randrange(850)+75
            rect = pygame.Rect(x-5,y-5,10,10)
            hit = False
            for w in walls:
                try:
                    s = (w[0][1]-w[1][1])/(w[0][0]-w[1][0])
                except:
                    s = 10000
                if s == 0:
                    s = 0.000001
                if ((s * rect.x + w[0][1] >= rect.y and s * rect.x + w[0][1] <= rect.y+rect.h)
                    or (s * (rect.x+rect.w) + w[0][1] >= rect.y and s * (rect.x+rect.w) + w[0][1] <= rect.y+rect.h)
                    or (rect.y/s >= rect.x and rect.y/s <= rect.x + rect.w)
                    or ((rect.y+rect.h)/s >= rect.x and (rect.y+rect.h) <= rect.x+rect.w)
                ):
                    hit = True
            if not hit:
                good = True
        self.x = x
        self.y = y
        self.rect = rect





    def getFitness(self,frames):
        return frames**2 - self.loss


    def upX(self,screen):
        self.x+= self.speed
        self.rect.move_ip(self.speed,0)
        hit = False
        wDistances = distances(self.walls,self,screen,self.visRad,False)
        hit = False
        for l in wDistances:
            if l < 5:
                hit = True

        if hit:
            self.x-= self.speed
            self.rect.move_ip(-self.speed,0)

    def downX(self,screen):
        self.x-= self.speed
        self.rect.move_ip(-self.speed,0)
        hit = False
        wDistances = distances(self.walls,self,screen,self.visRad,False)
        hit = False
        for l in wDistances:
            if l < 5:
                hit = True

        if hit:
            self.x+= self.speed
            self.rect.move_ip(+self.speed,0)

    def upY(self,screen):
        self.y+= self.speed
        self.rect.move_ip(0,self.speed)
        wDistances = distances(self.walls,self,screen,self.visRad,False)
        hit = False
        for l in wDistances:
            if l < 5:
                hit = True

        if hit:
            self.y-= self.speed
            self.rect.move_ip(0,-self.speed)

    def downY(self,screen):
        self.y-= self.speed
        self.rect.move_ip(0,-self.speed)
        hit = False
        wDistances = distances(self.walls,self,screen,self.visRad,False)
        hit = False
        for l in wDistances:
            if l < 5:
                hit = True

        if hit:
            self.x+= self.speed
            self.rect.move_ip(0,self.speed)

    def Hole(self):
        if self.isInHole == 0:
            distance = math.sqrt((self.x-self.closestHoleX)**2 + (self.y-self.closestHoleY)**2)
            if distance < 20:
                self.x = self.closestHoleX
                self.y = self.closestHoleY
                self.isInHole = 1
                self.rect.x = self.closestHoleX - self.rect.w/2
                self.rect.y = self.closestHoleY - self.rect.h/2
        else:
            self.isInHole = 0

    def capture(self):
        self.captured = True
        self.x = -100
        self.y = -100
        self.rect.x = -100
        self.rect.y = -100


    def getInputs(self,players,hunters,walls,holes,screen):

        d = 5000
        for h in holes:
            distance = math.sqrt((self.x-(h.x+h.w/2))**2 + (self.y-(h.x+h.h/2))**2)
            if distance < d:
                d = distance
                self.closestHoleX = h.x+h.w/2
                self.closestHoleY = h.y+h.h/2

        nnInputArray = [self.x/1000,self.y/1000,self.closestHoleX/1000,self.closestHoleY/1000]

        for h in hunters:
            nnInputArray.append(h.x/1000)
            nnInputArray.append(h.y/1000)
        for p in players:
            if p.x == self.x and p.y == self.y:
                continue
            nnInputArray.append(p.x/1000)
            nnInputArray.append(p.y/1000)

        nnInputArray.append(self.isInHole)


        wDistances = distances(walls,self,screen,self.visRad)

        for i in range(len(wDistances)):

            wDistances[i] = wDistances[i]/self.visRad

        nnInputArray.extend(wDistances)

        return nnInputArray







#######################################################
#######################################################
###################HUNTER CLASS########################
#######################################################
#######################################################





class Hunter:

    prevX = 1001
    prevY = 1001
    fitness = 0
    x = 0
    y = 0

    closestHoleX = 0
    closestHoleY = 0

    wallTraces = [0] * 16
    playerTraces = [0] * 16

    walls = None

    speed = 5

    checkLength = 0

    visRad = 100

    checkingHole = 0

    def __init__(self,walls,hunterCount,playerCount,seed):
        random.seed(seed)
        self.walls = walls
        x = 0
        y = 0
        rect = None
        good = False
        while not good:
            x = random.randrange(850)+75
            y = random.randrange(850)+75
            rect = pygame.Rect(x-5,y-5,10,10)
            hit = False
            for w in walls:
                if rect.colliderect(w):
                    hit = True
            if not hit:
                good = True
        self.x = x
        self.y = y
        self.rect = rect

    def getFitness(self,times):
        prevTime = 0
        for i,t in enumerate(times):
            self.fitness += 5000/(t-prevTime-i)
            prevTime = t


        if times[len(times)-1] < 1500:
            self.fitness += 500

        return self.fitness

    def upX(self,screen):
        self.x+= self.speed
        self.rect.move_ip(self.speed,0)
        wDistances = distances(self.walls,self,screen,self.visRad,False)
        hit = False
        for l in wDistances:
            if l < 5:
                hit = True

        if hit:
            self.x-= self.speed
            self.rect.move_ip(-self.speed,0)

    def downX(self,screen):
        self.x-= self.speed
        self.rect.move_ip(-self.speed,0)
        hit = False
        wDistances = distances(self.walls,self,screen,self.visRad,False)
        hit = False
        for l in wDistances:
            if l < 5:
                hit = True

        if hit:
            self.x+= self.speed
            self.rect.move_ip(self.speed,0)

    def upY(self,screen):
        self.y+= self.speed
        self.rect.move_ip(0,self.speed)
        hit = False
        wDistances = distances(self.walls,self,screen,self.visRad,False)
        hit = False
        for l in wDistances:
            if l < 5:
                hit = True

        if hit:
            self.y-= self.speed
            self.rect.move_ip(0,-self.speed)

    def downY(self,screen):
        self.y-= self.speed
        self.rect.move_ip(0,-self.speed)
        hit = False
        wDistances = distances(self.walls,self,screen,self.visRad,False)
        hit = False
        for l in wDistances:
            if l < 5:
                hit = True

        if hit:
            self.y+= self.speed
            self.rect.move_ip(0,self.speed)

    def Hole(self):
        distance = math.sqrt((self.x-self.closestHoleX)**2 + (self.y-self.closestHoleY)**2)
        if distance < 20:
            self.checkingHole = True


    def getInputs(self, players,walls,holes,screen):
        #punishes staying still
        if self.prevX == self.x and self.prevY == self.y:
            self.fitness += -10
        self.prevX = self.x
        self.prevY = self.y
        d = 5000
        for h in holes:
            distance = math.sqrt((self.x-(h.x+h.w/2))**2 + (self.y-(h.x+h.h/2))**2)
            if distance < d:
                d = distance
                self.closestHoleX = h.x+h.w/2
                self.closestHoleY = h.y+h.h/2


        nnInputArray = [self.x/1000,self.y/1000,self.closestHoleX/1000,self.closestHoleY/1000]

        wDistances = distances(walls,self,screen,self.visRad)

        playerLines = []
        for p in players:
            if p.isInHole == 0:
                playerLines.append([(p.rect.x,p.rect.y),(p.rect.x+p.rect.w,p.rect.y)])
                playerLines.append([(p.rect.x,p.rect.y+p.rect.h),(p.rect.x+p.rect.w,p.rect.y+p.rect.h)])
                playerLines.append([(p.rect.x,p.rect.y),(p.rect.x,p.rect.y+p.rect.h)])
                playerLines.append([(p.rect.x+p.rect.w,p.rect.y),(p.rect.x+p.rect.w,p.rect.y+p.rect.h)])
        pDistances = distances(playerLines,self,screen, self.visRad)

        for i in range (len(pDistances)):
            if wDistances[i] < pDistances[i]:
                pDistances[i] == self.visRad

        for d in range(len(pDistances)):
            #tiny incentive to get closer to players so that the hunter can more easily stumble into points
            if pDistances[d] != self.visRad:
                self.fitness += 1-pDistances[d]/self.visRad
                #bumps down player score a little bit so that they try and avoid the hunter
                #too lazy to change the code to return which player it is so I'll just bump
                #them all down and hope its helpful. it might make them coordinate? ¯\_(ツ)_/¯
                for p in players:
                    p.loss += 1-pDistances[d]/self.visRad

        for p in players:
            d = distances([[(p.rect.x,p.rect.y),(p.rect.x+p.rect.w,p.rect.y)],
                [(p.rect.x,p.rect.y+p.rect.h),(p.rect.x+p.rect.w,p.rect.y+p.rect.h)]
                ,[(p.rect.x,p.rect.y),(p.rect.x,p.rect.y+p.rect.h)],
                [(p.rect.x+p.rect.w,p.rect.y),(p.rect.x+p.rect.w,p.rect.y+p.rect.h)]]
                ,self,screen,self.visRad)
            for i in range(len(d)):
                if wDistances[i] < d[i]:
                    d[i] = self.visRad

            for n in d:
                if n < 8:
                    p.capture()


        for i in range(len(wDistances)):

            wDistances[i] = wDistances[i]/self.visRad

        for i in range(len(pDistances)):

            pDistances[i] = pDistances[i]/self.visRad


        nnInputArray.extend(wDistances)
        nnInputArray.extend(pDistances)
        return nnInputArray


raycaster.py
import math
import pygame

def distances(objects, entity,screen,rad, display = True):


    #32 rays, 11.25 degrees between them
    s = [10000, 5.0273, 2.4142,
    1.4966, 1, 0.6682,
    0.4142, 0.1989, 0.000000001,
    -0.1989, -0.4142, -0.6682,
    -1, -1.4966, -2.4142,
    -5.0273, -10000, 5.0273,
    2.4142, 1.4966, 1,
    0.6682, 0.4142, 0.1989,
    -0.000000001, -0.1989, -0.4142,
    -0.6682, -1, -1.4966,
    -2.4142,-5.0273]

    touching = [rad] * 32

    for o in objects:
        for i in range (0,len(s)):

            visRadPoint = (math.cos(math.atan(s[i]))*entity.visRad+entity.x,math.sin(math.atan(s[i]))*entity.visRad+entity.y)

            #visradPoint is b, entity is a

            d = visRadPoint[1] - entity.y
            e = entity.x - visRadPoint[0]
            f = d*entity.x + e*entity.y

            #startpoint is C, endpoint is D
            a = o[1][1] - o[0][1]
            b = o[0][0] - o[1][0]
            c = a*o[0][0] + b*o[0][1]

            determinant = d*b - a*e

            if determinant == 0:
                continue

            x = (b*f - e*c)/determinant
            y = (d*c - a*f)/determinant
            intersect = (x,y)

            if o[0][0] < o[1][0]:
                if intersect[0] < o[0][0] or intersect[0] > o[1][0]:
                    continue

            elif o[0][0] > o[1][0]:
                if intersect[0] > o[0][0] or intersect[0] < o[1][0]:
                    continue
            if o[0][1] < o[1][1]:
                if intersect[1] < o[0][1] or intersect[1] > o[1][1]:
                    continue
            else:
                if intersect[1] > o[0][1] or intersect[1] < o[1][1]:
                    continue

            distance = math.sqrt((intersect[0]-entity.x)**2 + (intersect[1]-entity.y)**2)

            if distance <= entity.visRad and distance < touching[i]:
                if i <= 16 and intersect[0] - entity.x > 0:
                    touching[i] = distance
                elif i > 16 and intersect[0] - entity.x < 0:
                    touching[i] = distance








    if display:
        for i in range(len(touching)):

            m = 1

            if i > 16:
                m = -1

            a = math.atan(s[i])

            pygame.draw.line(screen, (237, 122, 255), (entity.x, entity.y),(entity.x+m*int(math.cos(a)*entity.visRad),entity.y+m*int(math.sin(a)*entity.visRad)),1)

            if touching[i] != rad:
                pygame.draw.circle(screen,(252, 186, 3),(int(m*math.cos(a)*touching[i]+entity.x), int(m*math.sin(a)*touching[i]+entity.y)),4)


    return touching


tag.py
#for running individual simulations

import pygame
import enviornment
from player import *
from nnwk import NeuralNetwork
import random



class Sim:
    def __init__(self,playerCount,hunterCount,pShuffleRate,hShuffleRate,display,fLen):
        pygame.init()

        self.fLen = fLen
        self.display=display

        self.screen = pygame.display.set_mode([1000, 1000])

        self.running = True


        #loading enviornment
        self.walls = enviornment.walls()
        self.holes = []
        for i in enviornment.holes():
            self.holes.append(pygame.Rect(i[0],i[1],50,50))

        pygame.display.set_caption("test env")

        self.players = []
        self.hunters = []

        self.playerCount = playerCount
        self.hunterCount = hunterCount


        random.seed(os.urandom(5000))

        for i in range (0, self.playerCount):
            self.players.append(Player(self.walls,self.hunterCount,self.playerCount,random.randrange(200000)))
        for i in range (0, self.hunterCount):
            self.hunters.append(Hunter(self.walls,self.hunterCount,self.playerCount,random.randrange(200000)))

        #creating networks for hunters and players

        #player network will have 37 input nodes + those from players and hunters
        # 32 traces for walls
        # x pos, y pos, closestHoleX, closestHoleY, isInHole
        self.playerNetwork = NeuralNetwork(self.playerCount,self.hunterCount,37+2*(self.playerCount-1+self.hunterCount), 'player')
        self.playerNetwork.shuffle(1)
        #hunter network will have 68 input nodes
        # x pos, y pos, closestHoleX, closestHoleY
        # 32 traces for players
        # 32 traces for walls
        self.hunterNetwork = NeuralNetwork(self.playerCount,self.hunterCount,68, 'hunter')
        self.hunterNetwork.shuffle(1)


        self.frameCount = 0

        self.hCaptureTimes = [10000] * len(self.players)
        for i in range(len(self.hCaptureTimes)):
            self.hCaptureTimes[i] *= i+1

        pygame.font.init()
        self.font = pygame.font.SysFont('Arial', 14)


        #if in pure sim mode do not display
        if not self.display:
            os.putenv('SDL_VIDEODRIVER', 'fbcon')
            os.environ["SDL_VIDEODRIVER"] = "dummy"

    def runSim(self):

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                self.running = False






        self.screen.fill((255,255,255))

        pDecisions = []
        hDecisions = []




        #display players and hunters. separated for visuals
        for p in self.players:
            pygame.draw.circle(self.screen,(150, 194, 147),(int(p.x),int(p.y)), int(p.visRad))

        for h in self.hunters:
            pygame.draw.circle(self.screen,(255,100,100),(h.x,h.y),int(h.visRad))

        for h in self.holes:
            pygame.draw.rect(self.screen,(100,100,255),h)



        #display walls last so that they go above everything
        for w in self.walls:
            pygame.draw.line(self.screen,(0,0,0),w[0],w[1],4)

        for p in self.players:
            if not p.captured:
                pDecisions.append(self.playerNetwork.think(p.getInputs(self.players,self.hunters,self.walls,self.holes,self.screen),p))
            else:
                pDecisions.append([0,0,0,0])
        for h in self.hunters:
            if h.checkingHole == 0:
                hDecisions.append(self.hunterNetwork.think(h.getInputs(self.players,self.walls,self.holes,self.screen),h))
            else:
                hDecisions.append([0,0,0,0])


        for p in self.players:
            if p.isInHole == 0:
                pygame.draw.rect(self.screen,(0,255,0), p.rect)
            else:
                pygame.draw.rect(self.screen,(0,100,0), p.rect)
        for h in self.hunters:
            if not h.checkingHole:
                pygame.draw.rect(self.screen, (255,0,0), h.rect)
            else:
                pygame.draw.rect(self.screen, (100,0,0), h.rect)

        pygame.draw.rect(self.screen,(89, 0, 10),self.hunters[0].rect)



        for i in range(len(pDecisions)):
            #runs network's decided functions
            if self.players[i].captured == False:
                if pDecisions[i][0] and self.players[i].isInHole == 0:
                    self.players[i].upX(self.screen)
                    #print("upX")
                if pDecisions[i][1] and self.players[i].isInHole == 0:
                    self.players[i].downX(self.screen)
                    #print("downX")
                if pDecisions[i][2] and self.players[i].isInHole == 0:
                    self.players[i].upY(self.screen)
                    #print("upY")
                if pDecisions[i][3] and self.players[i].isInHole == 0:
                    self.players[i].downY(self.screen)
                    #print("downY")
                if pDecisions[i][4]:
                    self.players[i].Hole()
                #print("hole")
            #print("cycle")



        #display red rects as default:
        pygame.draw.rect(self.screen, (255,0,0), (975,900,12,12))
        pygame.draw.rect(self.screen, (255,0,0), (975,915,12,12))
        pygame.draw.rect(self.screen, (255,0,0), (975,930,12,12))
        pygame.draw.rect(self.screen, (255,0,0), (975,945,12,12))
        pygame.draw.rect(self.screen, (255,0,0), (975,960,12,12))


        if hDecisions[0][0]:
            pygame.draw.rect(self.screen, (0,255,0), (975,900,12,12))
        if hDecisions[0][1]:
            pygame.draw.rect(self.screen, (0,255,0), (975,915,12,12))
        if hDecisions[0][2]:
            pygame.draw.rect(self.screen, (0,255,0), (975,930,12,12))
        if hDecisions[0][3]:
            pygame.draw.rect(self.screen, (0,255,0), (975,945,12,12))
        try:
            if hDecisions[0][4]:
                pygame.draw.rect(self.screen, (0,255,0), (975,960,12,12))
        except:
            pass

        for i in range(len(hDecisions)):

            if hDecisions[i][0]:
                self.hunters[i].upX(self.screen)
            if hDecisions[i][1]:
                self.hunters[i].downX(self.screen)
            if hDecisions[i][2]:
                self.hunters[i].upY(self.screen)
            if hDecisions[i][3]:
                self.hunters[i].downY(self.screen)
            try:
                if hDecisions[i][4]:
                    self.hunters[i].Hole()
                    self.hunters[i].checkLength = 0
            except:
                pass

            if self.hunters[i].checkingHole:
                if self.hunters[i].checkLength > 30:
                    for p in self.players:
                        if p.isInHole:
                            if p.x == h.closestHoleX and p.y == h.closestHoleX:
                                for i in range(len(self.hCaptureTimes)):
                                    if self.hCaptureTimes[i] == 100000:
                                        self.hCaptureTimes[i] == self.frameCount
                                        break
                            p.capture()
                            p.fitness = p.getFitness(self.frameCount)

                else:
                    self.hunters[i].checkLength+=1

        if self.hCaptureTimes[len(self.hCaptureTimes)-1] == 100000 and self.frameCount<self.fLen:
            self.frameCount+=1
            text = self.font.render(str(self.frameCount), False, (0,32,107))
            self.screen.blit(text, (950,950))
        else:
            self.running = False

        if self.display:
            pygame.display.flip()


    def getResults(self):
        hFitness = self.hunters[0].getFitness(self.hCaptureTimes)
        #print("hf" + str(hFitness))

        worstPFitness = -100000000

        for p in self.players:
            if p.fitness > worstPFitness:
                worstPFitness = p.getFitness(self.frameCount)

        statement = []

        statement.append(hFitness)
        statement.append(self.hunterNetwork)
        statement.append(worstPFitness)
        statement.append(self.playerNetwork)


        return statement



        pygame.quit()
